sympy/polys/euclidtools.py

```diff
@@ -1880,18 +1880,35 @@ def dmp_cancel(f, g, u, K, include=True):
 
         cq, f = dmp_clear_denoms(f, u, K0, K, convert=True)
         cp, g = dmp_clear_denoms(g, u, K0, K, convert=True)
+
     else:
         cp, cq = K.one, K.one
 
     _, p, q = dmp_inner_gcd(f, g, u, K)
 
+    domul = True
     if K0 is not None:
-        _, cp, cq = K.cofactors(cp, cq)
+        _, _cp, _cq = K.cofactors(cp, cq)
 
         p = dmp_convert(p, u, K, K0)
         q = dmp_convert(q, u, K, K0)
-
-        K = K0
+        if (_cp, _cq).count(1) == 1 and 'QQ' in str(K0) and str(K) == 'ZZ':  # XXX what's the better way to test for this situation?
+            domul=False
+            def zz(l, c):
+                from mpmath.libmp import MPZ
+                rv = []
+                for i in l:
+                    if type(i) is  list:
+                        rv.append(zz(i))
+                    else:
+                        rv.append(MPZ(i*c))
+                return rv
+            p = zz(p, _cp)
+            q = zz(q, _cq)
```

Comment:
I don't know exactly what this is doing but we definitely don't want this sort of thing to be happening in this routine.

Reply:
It can be done more elegantly, but I concur with @Breydan-Smith that this is the change that needs to take place.

Reply:
It isn't just about elegance. This is not the right way to handle this at all.

What is needed is some reconsideration of exactly how to handle clearing denominators in the field case. It should not be special cased for ZZ and QQ because the relationship between those is the same as the relationship between e.g. `QQ[x]` and `QQ(x)`.

Reply:
The added code handles `QQ` and `QQ[x]`, doesn't it?